Kuadrant/kuadrant-operator
# enhancement: report dependency not installed to kuadrant & policy status

# Description
Closes: https://github.com/Kuadrant/kuadrant-operator/issues/730

Report in the Kuadrant and Policy status when required dependencies are not installed

* Kuadrant CR
  * Gateway API not installed
  * Gateway Provider (istio / envoy gateway) not intalled
  * Limitador Operator not installed
  * Authorino Operator not installed
* AuthPolicy CR
  * Gateway API not installed
  * Gateway Provider (istio / envoy gateway) not intalled
  * Authorino Operator not installed
* RateLimitPolicy CR
  * Gateway API not installed
  * Gateway Provider (istio / envoy gateway) not intalled
  * Limitador Operator not installed
* TLSPolicy CR (does **not**  require a Gateway Provider)
  * Gateway API not installed
  * CertManager not installed
* DNSPolicy CR (does **not**  require a Gateway Provider)
  * Gateway API not installed
  * DNS Operator not installed

# Verification
* Gateway API and Gateway API provider status are tested by integration tests, so passing integration tests should be enough

To verify other status:
* Create cluster
```sh
make local-env-setup
```
* Delete CRDs to simulate not installed
```sh
# DNS Operator
kubectl delete crd dnsrecords.kuadrant.io   
# Authorino Operator                                                                    
kubectl delete crd authconfigs.authorino.kuadrant.io
kubectl delete crd authorinos.operator.authorino.kuadrant.io
# Limitador Operator
kubectl delete crd limitadors.limitador.kuadrant.io
# Cert Manager
kubectl delete crd certificates.cert-manager.io
```
* Run operator
```
make run
```
* Apply CRs
```sh
kubectl apply -f config/samples/kuadrant_v1_authpolicy.yaml                                      
kubectl apply -f config/samples/kuadrant_v1_dnspolicy.yaml
kubectl apply -f config/samples/kuadrant_v1_ratelimitpolicy.yaml
kubectl apply -f config/samples/kuadrant_v1_tlspolicy.yaml
kubectl apply -f config/samples/kuadrant_v1beta1_kuadrant.yaml
```
* Check status
```sh
kubectl get kuadrant kuadrant-sample -o yaml | yq '.status'
kubectl get authpolicy authpolicy-sample -o yaml | yq '.status'
kubectl get ratelimitpolicy ratelimitpolicy-sample -o yaml | yq '.status'
kubectl get dnspolicy dnspolicy-sample -o yaml | yq '.status'
kubectl get tlspolicy tlspolicy-sample -o yaml | yq '.status'
```

![image](https://github.com/user-attachments/assets/bb1143c1-f7b8-443f-b32c-7cd07312fa1b)
